src/metabase/db/liquibase.clj

```diff
@@ -228,14 +232,55 @@
   [liquibase]
   (.hasChangeLogLock (lock-service liquibase)))
 
+(defn- wait-until [done? ^long sleep-ms timeout-ms]
+  (let [deadline (+ (System/nanoTime) (* 1e6 timeout-ms))]
+    (loop []
+      (if (done?)
+        :done
+        (if (>= (System/nanoTime) deadline)
+          :timed-out
+          (do (Thread/sleep sleep-ms)
+              (recur)))))))
+
+(defn- locked-instances []
+  (filter holding-lock? @potentially-locked-instances))
+
+(defn wait-for-all-locks
+  "Wait up to a maximum of `timeout-seconds` for the given Liquibase instance to release the migration lock."
+  [sleep-ms timeout-ms]
+  (log/warn (locked-instances))
+  (let [done? #(empty? (locked-instances))]
+    (if (done?)
+      :none
+      (do (log/infof "Waiting for migration lock(s) to be released (max %.1f secs)" (/ timeout-ms 1000))
+          (wait-until done? sleep-ms timeout-ms)))))
+
+(defn- liquibase->url [^Liquibase liquibase]
+  ;; Need to this cast to get access to the metadata. We currently only use JDBC app databases.
+  (let [conn ^JdbcConnection (.. liquibase getDatabase getConnection)]
+    (-> conn .getMetaData .getURL)))
+
+(defn release-concurrent-locks!
+  "Release any locks held by this process corresponding to the same database."
+  [conn-or-data-source]
+  (when-let [instances (not-empty (locked-instances))]
+    ;; We cannot use the existing instances to clear the locks, as their connections are currently blocked.
+    ;; Since we cannot "clone" a connection (they have "forgotten" their password), we use this new conn.
+    (with-liquibase [liquibase conn-or-data-source]
+      (let [url (liquibase->url liquibase)]
+        (doseq [instance instances]
+          (when (= url (liquibase->url instance))
+            (log/warnf "Releasing liquibase lock on %s before migrations finished" url)
```

Comment:
we shouldn't log the url in any case since it might contains sensitive info.

Reply:
The password will always be stripped but I guess the username could be sensitive, will remove.